patches/tModLoader/Terraria/GameContent/ShopHelper.cs.patch

```diff
@@ -69,26 +69,28 @@
  			new AllPersonalitiesModifier().ModifyShopPrice(info, this);
  			if (_currentHappiness == "")
  				AddHappinessReportText("Content");
-@@ -167,11 +_,22 @@
+@@ -167,11 +_,23 @@
  		return priceAdjustment;
  	}
  
+-	private static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue("TownNPCMoodBiomes." + biomeNameKey);
 +	//TML: Turned public.
 +	//TODO: Anything else?
--	private static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue("TownNPCMoodBiomes." + biomeNameKey);
-+	public static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue("TownNPCMoodBiomes." + biomeNameKey);
++	public static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue(biomeNameKey.StartsWith("Mods.") ? biomeNameKey : "TownNPCMoodBiomes." + biomeNameKey);
  
- 	private void AddHappinessReportText(string textKeyInCategory, object substitutes = null)
+-	private void AddHappinessReportText(string textKeyInCategory, object substitutes = null)
++	private void AddHappinessReportText(string textKeyInCategory, object substitutes = null, int npcType = 0)
  	{
 +		string text;
 +
-+		if (_currentNPCBeingTalkedTo.type < NPCID.Count)
++		if(textKeyInCategory == "Princess_LovesNPC")
```

Comment:
Something to think about in the future, is there any way for mods to add these sort of relationships between their own NPCs and another mod's?

Feature 1: `ModTownNPC` from another mod adds a special localization for liking a specific NPC, rather than just the general substitution based one.

Feature 2: A way for another mod to make `ExamplePerson` like `ModTownNPC`, with the localization defined in `ModTownNPC` (like `ExamplePerson_LovesNPC`)

Feel free to convert to a feature request issue

Reply:
I haven't seen any modder care enough about this to mention it. Mostly modders have been confused that princess even has those npc specific texts. I'll make an issue though.

Reply:
https://github.com/tModLoader/tModLoader/issues/3493